File: mobile/src/main/java/treehou/se/habit/ui/control/CellFactory.java
package treehou.se.habit.ui.control;

import android.content.Context;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.LinearLayout;
import android.widget.RemoteViews;

import java.util.HashMap;
import java.util.Map;

import treehou.se.habit.R;
import treehou.se.habit.core.db.model.controller.CellDB;
import treehou.se.habit.core.db.model.controller.ControllerDB;

public class CellFactory<T> {

    private static final String TAG = "CellFactory";

    private Map<T, CellBuilder> cellBuilders = new HashMap<>();
    private CellBuilder defaultBuilder = new DefaultBuilder();

    public void addBuilder(T type, CellBuilder builder){
        cellBuilders.put(type, builder);
    }

    public void setDefaultBuilder(CellBuilder builder){
        defaultBuilder = builder;
    }

    public View create(Context context, ControllerDB controller, final CellDB cell){

        Log.d(TAG,"cellBuilder cell type " + cell.getType());
        View cellView;
        try {
            int cellType = cell.getType();

            CellBuilder cellBuilder = cellBuilders.get(cellType);
            if (cellBuilder == null) {
                Log.d(TAG, "cellBuilder using default");
                cellBuilder = defaultBuilder;
            } else {
                Log.d(TAG, "cellBuilder using custom");
            }
            cellView = cellBuilder.build(context, controller, cell);
        }catch (Exception e){
            Log.d(TAG, "Failed render " + cell + " error " + e);
            e.printStackTrace();
            cellView = defaultBuilder.build(context, controller, cell);
        }

        LinearLayout.LayoutParams lp = new LinearLayout.LayoutParams(
                LinearLayout.LayoutParams.MATCH_PARENT, LinearLayout.LayoutParams.MATCH_PARENT);
        lp.weight = 1;
        cellView.setLayoutParams(lp);

        return cellView;
    }

    public RemoteViews createRemote(Context context, ControllerDB controller, final CellDB cell){

        Log.d(TAG,"cellBuilder cell type " + cell.getType());

        CellBuilder cellBuilder = cellBuilders.get(cell.getType());

        if(cellBuilder == null) {
            Log.d(TAG,"cellBuilder using default");
            cellBuilder  = defaultBuilder;
        }else{
            Log.d(TAG,"cellBuilder using custom");
        }

        RemoteViews remoteViews;
        try {
            remoteViews = cellBuilder.buildRemote(context, controller, cell);
        }catch (Exception e){
            remoteViews = defaultBuilder.buildRemote(context, controller, cell);
        }

        return remoteViews;
    }

    public interface CellBuilder {

        View build(Context context, ControllerDB controller, CellDB cell);
        RemoteViews buildRemote(Context context, ControllerDB controller, CellDB cell);
    }

    public static class DefaultBuilder implements CellBuilder {

        public View build(Context context, ControllerDB controller, CellDB cell){

            LayoutInflater inflater = LayoutInflater.from(context);
            return inflater.inflate(R.layout.cell_empty, null);
        }

        @Override
        public RemoteViews buildRemote(Context context, ControllerDB controller, CellDB cell) {
            return new RemoteViews(context.getPackageName(), R.layout.cell_empty);
        }
    }
}

File: mobile/src/main/java/treehou/se/habit/ui/control/CellFactory.kt
package treehou.se.habit.ui.control

import android.content.Context
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.LinearLayout
import android.widget.RemoteViews

import java.util.HashMap

import treehou.se.habit.R
import treehou.se.habit.core.db.model.controller.CellDB
import treehou.se.habit.core.db.model.controller.ControllerDB

class CellFactory {

    private val cellBuilders = HashMap<Int, CellBuilder>()
    private var defaultBuilder: CellBuilder = DefaultBuilder()

    fun addBuilder(type: Int, builder: CellBuilder) {
        cellBuilders.put(type, builder)
    }

    fun setDefaultBuilder(builder: CellBuilder) {
        defaultBuilder = builder
    }

    fun create(context: Context, container: ViewGroup, controller: ControllerDB, cell: CellDB): View {

        Log.d(TAG, "cellBuilder cell type " + cell.type)
        var cellView: View
        try {
            val cellType = cell.type

            var cellBuilder: CellBuilder? = cellBuilders[cellType]
            if (cellBuilder == null) {
                Log.d(TAG, "cellBuilder using default")
                cellBuilder = defaultBuilder
            } else {
                Log.d(TAG, "cellBuilder using custom")
            }
            cellView = cellBuilder.build(context, container, controller, cell)
        } catch (e: Exception) {
            Log.d(TAG, "Failed render $cell error $e")
            e.printStackTrace()
            cellView = defaultBuilder.build(context, container, controller, cell)
        }

        val lp = LinearLayout.LayoutParams(
                LinearLayout.LayoutParams.MATCH_PARENT, LinearLayout.LayoutParams.MATCH_PARENT)
        lp.weight = 1f
        cellView.layoutParams = lp

        return cellView
    }

    fun createRemote(context: Context, controller: ControllerDB, cell: CellDB): RemoteViews? {

        Log.d(TAG, "cellBuilder cell type " + cell.type)

        var cellBuilder: CellBuilder? = cellBuilders[cell.type]

        if (cellBuilder == null) {
            Log.d(TAG, "cellBuilder using default")
            cellBuilder = defaultBuilder
        } else {
            Log.d(TAG, "cellBuilder using custom")
        }

        var remoteViews: RemoteViews?
        try {
            remoteViews = cellBuilder.buildRemote(context, controller, cell)
        } catch (e: Exception) {
            remoteViews = defaultBuilder.buildRemote(context, controller, cell)
        }

        return remoteViews
    }

    interface CellBuilder {

        fun build(context: Context, container: ViewGroup, controller: ControllerDB, cell: CellDB): View
        fun buildRemote(context: Context, controller: ControllerDB, cell: CellDB): RemoteViews?
    }

    class DefaultBuilder : CellBuilder {

        override fun build(context: Context, container: ViewGroup, controller: ControllerDB, cell: CellDB): View {

            val inflater = LayoutInflater.from(context)
            return inflater.inflate(R.layout.cell_empty, container, false)
        }

        override fun buildRemote(context: Context, controller: ControllerDB, cell: CellDB): RemoteViews {
            return RemoteViews(context.packageName, R.layout.cell_empty)
        }
    }

    companion object {

        private val TAG = "CellFactory"
    }
}
